Libraries/MiscDrivers/EEPROM/eeprom_24lc256_driver.c

```diff
@@ -0,0 +1,239 @@
+static int i2c_transfer(mxc_i2c_req_t *req, uint8_t *txData, int txSize, uint8_t *rxData,
+                        int rxSize)
+{
+    req->tx_buf = txData; // Write data buffer
+    req->tx_len = txSize; // Number of bytes to write
+    req->rx_buf = rxData; // Read data buffer
+    req->rx_len = rxSize; // Number of bytes to read
+    return MXC_I2C_MasterTransaction(req);
+}
+
+static int i2c_write(mxc_i2c_req_t *req, uint8_t *txData, int txSize)
+{
+    return i2c_transfer(req, txData, txSize, NULL, 0); // Create I2C write request
+}
+
+static int i2c_read(mxc_i2c_req_t *req, uint8_t *txData, uint8_t *rxData, int rxSize)
+{
+    return i2c_transfer(req, txData, 2, rxData, rxSize); // Create I2C read request
+}
+
+/**
+ * @brief   Initializes I2C registers EEPROM
+ * @param   i2c		I2C registers
+ * @param   addr	Slave I2C address of EEPROM.
+ * @returns #E_NO_ERROR if init succeeded.
+ *
+ */
+static int eeprom_24LC256_init(mxc_i2c_regs_t *i2c, uint8_t addr)
+{
+    req.i2c = i2c;
+    req.addr = addr;
+    req.tx_buf = NULL;
+    req.tx_len = 0;
+    req.rx_buf = NULL;
+    req.rx_len = 0;
+    req.restart = 0;
+    req.callback = NULL;
+
+    return E_NO_ERROR;
+}
+
+/**
+ * @brief   Writes data from EEPROM
+ * @param   addr		Start address we want to read.
+ * @param   data_buffer	Data buffer to read.
+ * @param   length		Number of bytes to read.
+ * @returns #E_NO_ERROR if read succeeded. non-zero if an error occurred.
+ *
+ */
+static int eeprom_24LC256_read(uint16_t addr, uint8_t *data_buffer, uint16_t length)
+{
+    int err = E_NO_ERROR;
+    uint16_t remaining = length;
+    uint8_t *current_data_buffer = data_buffer;
+    uint16_t eeprom_addr_to_read = addr;
+    uint8_t send_buffer[2] = { 0x00, 0x00 };
+
+    if (addr + length > _24LC256_EEPROM_SIZE) {
+        return E_BAD_PARAM;
+    }
+
+    while (remaining) {
+        if (remaining <= I2C_MAX_READ_SIZE) {
+            send_buffer[0] = eeprom_addr_to_read >> 8;
+            send_buffer[1] = (eeprom_addr_to_read & 0x00FF);
+
+            // Read
+            err = i2c_read(&req, send_buffer, current_data_buffer, remaining);
+            if (err != E_NO_ERROR) {
+                return err;
+            } else {
+                remaining = 0;
+            }
+        } else {
+            send_buffer[0] = eeprom_addr_to_read >> 8;
+            send_buffer[1] = (eeprom_addr_to_read & 0x00FF);
+
+            // Read
+            err = i2c_read(&req, send_buffer, current_data_buffer, I2C_MAX_READ_SIZE);
+            if (err != E_NO_ERROR) {
+                return err;
+            }
+            remaining -= I2C_MAX_READ_SIZE;
+            current_data_buffer += I2C_MAX_READ_SIZE;
+            eeprom_addr_to_read += I2C_MAX_READ_SIZE;
+        }
+    }
+    return E_NO_ERROR;
+}
+
+/**
+ * @brief   Writes a small chunk of data directly to the EEPROM. The written memory should be in the same page of EEPROM (1 page = 64 bytes)
```

Comment:
Given this limitation I would suggest throwing an error code if the `length` parameter is too large.

Reply:
I added throwing error.